### PR TITLE
Feature/new executions

### PR DESCRIPTION
**Pull request by kuegi**
switched handling of executions (not only full order, but also partial fills)

also fixed backtestengine

### CODE REVIEW

File: kuegi_bot/backtest_engine.py [hunk 1]
```diff
@@ -175,13 +177,23 @@ def orderKeyForSort(self,order):
         else: # limit -> bigger numbers to be sorted after the stops
             return abs(self.current_bars[0].close - order.limit_price)+self.current_bars[0].close
 
-    def handle_open_orders(self, intrabarToCheck: Bar) -> bool:
-        something_changed = False
+    def check_executions(self, intrabarToCheck: Bar, onlyOnClose):
         another_round= True
+        didSomething= False
```
Comment (Krysh90) [inline kuegi_bot/backtest_engine.py]: rename to did_something

File: kuegi_bot/backtest_engine.py [hunk 2]
```diff
@@ -175,13 +177,23 @@ def orderKeyForSort(self,order):
...
+        allowedOrderIds= None
```
Comment (Krysh90) [inline kuegi_bot/backtest_engine.py]: allowed_order_ids

File: kuegi_bot/backtest_engine.py [hunk 3]
```diff
@@ -198,17 +210,32 @@ def handle_open_orders(self, intrabarToCheck: Bar) -> bool:
...
                 if should_execute:
                     self.handle_order_execution(order, intrabarToCheck)
                     self.bot.on_tick(self.current_bars, self.account)
                     another_round= True
-                    something_changed= True
+                    didSomething= True
                     break
+        return didSomething
+
+    def handle_subbar(self, intrabarToCheck: Bar):
```
Comment (Krysh90) [inline kuegi_bot/backtest_engine.py:260]: intrabar_to_check

File: kuegi_bot/backtest_engine.py [hunk 4]
```diff
@@ -198,17 +210,32 @@ def handle_open_orders(self, intrabarToCheck: Bar) -> bool:
...
+        # first the ones that are there at the beginning
+        orders= list(self.account.open_orders)
+        didSomething1= self.check_executions(intrabarToCheck,False)
+        self.current_bars[0].add_subbar(intrabarToCheck)  # so bot knows about the current intrabar
+
+        didSomething2= self.check_executions(intrabarToCheck,True)
```
Comment (Krysh90) [inline kuegi_bot/backtest_engine.py]: didSomething1 and 2? is there no more meaningful meaning for those?

anyways would be did_something_1 and did_something_2

File: kuegi_bot/bots/trading_bot.py [hunk 5]
```diff
@@ -39,7 +39,7 @@ def __init__(self, logger, directionFilter: int = 0):
         self.time_of_max_equity = 0
         self.position_history: List[Position] = []
         self.openPositionRolling = 1
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py]: open_position_rolling

File: kuegi_bot/bots/trading_bot.py [hunk 6]
```diff
@@ -39,7 +39,7 @@ def __init__(self, logger, directionFilter: int = 0):
...
-        self.unaccountedPositionCoolOff= 0
+        self.unaccountedPositionCoolOff = 0
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py]: unaccounted_position_cool_off

File: kuegi_bot/bots/trading_bot.py [hunk 7]
```diff
@@ -171,15 +171,61 @@ def position_got_opened(self, position: Position, bars: List[Bar], account: Acco
...
+    def handle_opened_position(self, position: Position, account: Account, bars: List[Bar]):
         position.status = PositionStatus.OPEN
-        position.filled_entry = order.executed_price if order is not None else None
-        position.entry_tstamp = order.execution_tstamp if order is not None and order.execution_tstamp > 0 else bars[
-            0].tstamp
-
         self.position_got_opened(position, bars, account)
 
+    def on_execution(self, orderId, amount, executed_price, tstamp):
+        posId = self.position_id_from_order_id(orderId)
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py]: pos_id

File: kuegi_bot/bots/trading_bot.py [hunk 8]
```diff
@@ -171,15 +171,61 @@ def position_got_opened(self, position: Position, bars: List[Bar], account: Acco
...
-    def handle_opened_position(self, position: Position, order: Order, account: Account, bars: List[Bar]):
+    def handle_opened_position(self, position: Position, account: Account, bars: List[Bar]):
         position.status = PositionStatus.OPEN
-        position.filled_entry = order.executed_price if order is not None else None
-        position.entry_tstamp = order.execution_tstamp if order is not None and order.execution_tstamp > 0 else bars[
-            0].tstamp
-
         self.position_got_opened(position, bars, account)
 
+    def on_execution(self, orderId, amount, executed_price, tstamp):
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py]: order_id

File: kuegi_bot/bots/trading_bot.py [hunk 9]
```diff
@@ -171,15 +171,61 @@ def position_got_opened(self, position: Position, bars: List[Bar], account: Acco
...
+        posId = self.position_id_from_order_id(orderId)
+        if posId not in self.open_positions.keys():
+            self.logger.info("executed order not found in positions: " + orderId)
+            return
+        position: Position = self.open_positions[posId]
+
+        orderType = self.order_type_from_order_id(orderId)
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py]: order_type

File: kuegi_bot/bots/trading_bot.py [hunk 10]
```diff
@@ -171,15 +171,61 @@ def position_got_opened(self, position: Position, bars: List[Bar], account: Acco
...
+        position.changed = True
+        if orderType == OrderType.ENTRY:
+            position.status = PositionStatus.OPEN
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py:189]: missing logger.info for position got opened?

File: kuegi_bot/bots/trading_bot.py [hunk 11]
```diff
@@ -188,26 +234,22 @@ def sync_executions(self, bars: List[Bar], account: Account):
...
-
-                elif (
-                        orderType == OrderType.SL or orderType == OrderType.TP) and position.status == PositionStatus.OPEN:
-                    self.logger.info("position %s got closed" % position.id)
-                    position.status = PositionStatus.CLOSED
-                    position.filled_exit = order.executed_price
-                    position.exit_tstamp = order.execution_tstamp
-                    position.exit_equity = account.equity
-                    self.position_closed(position, account)
-                else:
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py:208]: this removed part has been moved to on_execution - right?

File: kuegi_bot/bots/trading_bot.py [hunk 12]
```diff
@@ -443,7 +487,7 @@ def save_open_positions(self, bars: List[Bar]):
             with open(base + self._get_pos_file(self.openPositionRolling), 'w') as file:
                 file.write(string)
             self.openPositionRolling += 1
-            if self.openPositionRolling > 10:
+            if self.openPositionRolling > 3:
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py]: maybe have a constant somewhere, which is importable that everyone is able to define how much rolling position files they want to have?

File: kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py [hunk 13]
```diff
@@ -8,13 +8,40 @@
 
 class BybitLinearPublicPart(KuegiWebsocket):
 
-    def __init__(self, wsURLs, api_key, api_secret, logger, callback, on_message):
+    def __init__(self, wsURLs, api_key, api_secret, logger, callback, on_message,
+                 symbol, minutesPerBar):
         self.messageCallback= on_message
+        self.minutesPerBar= minutesPerBar
```
Comment (Krysh90) [inline kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py]: minutes_per_bar

File: kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py [hunk 14]
```diff
@@ -8,13 +8,40 @@
...
+        self.symbol= symbol
+        self.initialSubscribeDone= False
```
Comment (Krysh90) [inline kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py]: initial_subscribe_done

File: kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py [hunk 15]
```diff
@@ -8,13 +8,40 @@
...
         super().__init__(wsURLs, api_key, api_secret, logger, callback)
 
     def on_message(self, message):
         self.messageCallback(message)
 
+    def subscribeRealtimeData(self):
```
Comment (Krysh90) [inline kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py]: subscribe_realtime_data

File: kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py [hunk 16]
```diff
@@ -8,13 +8,40 @@
...
+        subbarsIntervall = '1' if self.minutesPerBar <= 60 else '60'
```
Comment (Krysh90) [inline kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py]: subbars_interval

File: kuegi_bot/utils/trading_classes.py [hunk 17]
```diff
@@ -182,14 +188,19 @@ def __init__(self, id: str, entry: float, stop: float, amount: float, tstamp):
         self.id: str = id
         self.signal_tstamp = tstamp
         self.status: PositionStatus = PositionStatus.PENDING
+        self.changed= False
         self.wanted_entry = entry
         self.initial_stop = stop
         self.amount = amount
+        self.maxFilledAmount= 0
```
Comment (Krysh90) [inline kuegi_bot/utils/trading_classes.py]: max_filled_amount

File: kuegi_bot/utils/trading_classes.py [hunk 18]
```diff
@@ -182,14 +188,19 @@ def __init__(self, id: str, entry: float, stop: float, amount: float, tstamp):
...
+        self.currentOpenAmount= 0
```
Comment (Krysh90) [inline kuegi_bot/utils/trading_classes.py]: current_open_amount

File: kuegi_bot/utils/trading_classes.py [hunk 19]
```diff
@@ -182,14 +188,19 @@ def __init__(self, id: str, entry: float, stop: float, amount: float, tstamp):
...
+        self.last_filled_entry: float= None
         self.filled_entry: float = None
         self.filled_exit: float = None
         self.entry_tstamp = 0
         self.exit_tstamp = 0
         self.exit_equity = 0
+        self.customData= {}
```
Comment (Krysh90) [inline kuegi_bot/utils/trading_classes.py]: custom_data

File: kuegi_bot/backtest_engine.py [hunk 20]
```diff
@@ -93,6 +94,14 @@ def send_order(self, order: Order):
         if order.amount == 0:
             self.logger.error("trying to send order without amount")
             return
+        [posId, order_type]= TradingBot.position_id_and_type_from_order_id(order.id)
+        if order_type == OrderType.ENTRY:
+            [posId, direction]= TradingBot.split_pos_Id(posId)
```
Comment (Krysh90) [inline kuegi_bot/backtest_engine.py]: shouldn't it be called `split_pos_id`?

File: kuegi_bot/backtest_engine.py [hunk 21]
```diff
@@ -170,45 +182,86 @@ def handle_order_execution(self, order: Order, intrabar: Bar):
     def orderKeyForSort(self,order):
         if order.stop_price is None and order.limit_price is None:
             return 0
+        # sort buys after sells (higher number) when bar is falling
+        longFac= 1 if self.bars[0].close > self.bars[0].open else 2
+        shortFac= 1 if self.bars[0].close < self.bars[0].open else 2
```
Comment (Krysh90) [inline kuegi_bot/backtest_engine.py]: long_fac and short_fac

File: kuegi_bot/backtest_engine.py [hunk 22]
```diff
@@ -170,45 +182,86 @@ def handle_order_execution(self, order: Order, intrabar: Bar):
...
-                        elif ((order.amount > 0 and order.limit_price > intrabarToCheck.close) or (
-                                order.amount < 0 and order.limit_price < intrabarToCheck.close)):
+                            if only_on_close: # order just came in and executed right away: execution on the worst price cause can't assume anything better
+                                order.stop_price= intrabar_to_check.low if order.amount < 0 else intrabar_to_check.high
+
+                        elif ((order.amount > 0 and order.limit_price > intrabar_to_check.close) or (
+                                order.amount < 0 and order.limit_price < intrabar_to_check.close)):
                             # close below/above limit: got definitly executed
                             should_execute= True
+                            force_taker= True # need to assume taker.
```
Comment (Krysh90) [inline kuegi_bot/backtest_engine.py]: this was one of those changes, which makes backtest engine more reliable in sense of live executions?

File: kuegi_bot/backtest_engine.py (hunk 3 above)
Comment (Krysh90) [reply kuegi_bot/backtest_engine.py:260]: resolved but not changed? 🤔  i mean yeah it is only code-styling 🤷

File: kuegi_bot/bots/MultiStrategyBot.py [hunk 23]
```diff
@@ -67,7 +62,7 @@ def manage_open_order(self, order, position, bars, to_update, to_cancel, open_po
     def manage_open_position(self, p, bars, account, pos_ids_to_cancel):
         pass
 
-    def open_orders(self, is_new_bar, directionFilter, bars, account, open_positions):
+    def open_orders(self, is_new_bar, directionFilter, bars, account, open_positions :dict):
```
Comment (Krysh90) [inline kuegi_bot/bots/MultiStrategyBot.py]: minor code-styling: whitespace on wrong side of :

File: kuegi_bot/bots/strategies/kuegi_strat.py [hunk 24]
```diff
@@ -34,29 +35,43 @@ def init(self, bars: List[Bar], account: Account, symbol: Symbol):
...
         if not gotStop:
             order = Order(orderId=TradingBot.generate_order_id(positionId=position.id,
                                                            type=OrderType.SL),
                       stop=position.initial_stop,
                       amount=-position.amount)
             self.order_interface.send_order(order)
-            # added temporarily, cause sync with open orders is in the next loop and otherwise the orders vs
-            # position check fails
-            if order not in account.open_orders:  # outside world might have already added it
-                account.open_orders.append(order)
```
Comment (Krysh90) [inline kuegi_bot/bots/strategies/kuegi_strat.py:59]: is this removal intended? maybe some unwanted change from PR #15 ?

File: kuegi_bot/bots/trading_bot.py [hunk 25]
```diff
@@ -26,20 +26,20 @@ def __init__(self, logger, directionFilter: int = 0):
...
-        self.order_interface: OrderInterface = None
+        self.order_interface: ExchangeInterface = None
         self.symbol: Symbol = None
         self.unique_id: str = ""
         self.last_time = 0
-        self.last_tick_time : datetime = None
+        self.last_tick_time: datetime = None
         self.is_new_bar = True
-        self.open_positions = {}
+        self.open_positions : Dict[str,Position] = {}
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py]: minor code-styling: to be consistent with other properties in here. no leading whitespace before :

File: kuegi_bot/bots/trading_bot.py [hunk 26]
```diff
@@ -167,50 +168,115 @@ def cancel_entry(self, positionId, account: Account):
...
+
+        for position in closed_pos:
+            self.position_closed(position, account)
+
+        if changed:
+            self.sync_connected_orders(account)
+
+        # old way (fallback if executions wheren't there)
+        if not self.order_interface.handles_executions:
+            self.logger.warn("your exchange interface doesn't handle executions yet. please upgrade")
```
Comment (Krysh90) [inline kuegi_bot/bots/trading_bot.py]: maybe this should be an error, to be more visible

File: kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py (hunk 13 above)
Comment (Krysh90) [reply kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py]: this property got changed in class below, but not in this?

File: kuegi_bot/backtest_engine.py (hunk 22 above)
Comment (kuegi) [reply kuegi_bot/backtest_engine.py]: yes

File: kuegi_bot/bots/strategies/kuegi_strat.py (hunk 24 above)
Comment (kuegi) [reply kuegi_bot/bots/strategies/kuegi_strat.py:59]: intented, as this safety check is now done in trade_engine directly. -> no need to do it in every strat anymore
